src/parser/expression.js

```diff
@@ -956,16 +1034,17 @@ export default class ExpressionParser extends LValParser {
 
   // Parse object or class method.
 
-  parseMethod(node: N.MethodLike, isGenerator?: boolean, isAsync?: boolean): N.MethodLike {
+  parseMethod<T : N.MethodLike>(node: T, isGenerator: boolean, isAsync: boolean, isConstructor: boolean, type: string): T {
     const oldInMethod = this.state.inMethod;
     this.state.inMethod = node.kind || true;
     this.initFunction(node, isAsync);
     this.expect(tt.parenL);
-    node.params = this.parseBindingList(tt.parenR);
+    const allowModifiers = isConstructor; // For TypeScript parameter properties
+    node.params = this.parseBindingList(tt.parenR, /* allowEmpty */ false, allowModifiers);
     node.generator = !!isGenerator;
-    this.parseFunctionBody(node);
+    const missingBody = this.maybeParseFunctionBody(node, /* allowExpression */ false, /* allowMissingBody */ type === "ClassMethod");
```

Comment:
I think we could pass an object here, like:

```js
maybeParseFunctionBody({ allowExpression: false, allowMissingBody: type === "ClassMethod"});
```

Looks better than comments to me.